designs/2021-fixable-disable-directives/design.md

```diff
@@ -0,0 +1,112 @@
+- `_verifyWithoutProcessors` is called within the call stack of each of the 1-10 passes in the linter's `verifyAndFix`
+
+This RFC proposes making three changes:
+
+- In the patched `options.fix`, consider a problem's meta type to be `"meta"` if its `ruleId` is `null`
+- Pass the `SourceCode` being linted as a parameter to `applyDisableDirectives`
+- Add a `fix` method to the unused directive problems returned by `applyDirectives` that uses the `SourceCode` to compute:
+  - For comments that are the only non-whitespace on their line, delete that line
+  - Otherwise, delete just the comment and any now-unnecessary surrounding whitespace
+
+### Fix Behavior Examples
+
+```diff
+- /* eslint-disable */
+```
+
+```diff
+- // eslint-disable-next-line -- related explanation
+```
+
+```diff
+- before /* eslint-disable-next-line -- related explanation */ after
++ before after
+```
+
+```diff
+- before // eslint-disable-next-line -- related explanation
++ before
+```
+
+```diff
+  // before
+- // eslint-disable-next-line
+  // after
+```
+
+Multiline block comments are already not allowed to be disable directives. [eslint/eslint#10334](https://github.com/eslint/eslint/issues/10334)
+
+## Documentation
+
+- [Command Line Interface > Fixing Problems](https://eslint.org/docs/user-guide/command-line-interface#fixing-problems)'s `--fix` documentation should mention the added directives fixing and the new `--fix-type`
+- [Rules > Report Unused ESLint Disable Comments](https://eslint.org/docs/user-guide/configuring/rules#report-unused-eslint-disable-comments) should link to that documentation
+
+## Drawbacks
+
+Like any new feature, this flag will slightly increase the complexity and maintenance costs of ESLint core.
+
+This RFC's implementation would lock in the name for a new `--fix-type` even though we only have one concrete use case for it.
+
+## Backwards Compatibility Analysis
+
+It is unlikely but not impossible that some `--fix` dependant users would rely on unused disable directives remaining in code.
+
+Otherwise, this change is additive in behavior.
+
+## Alternatives
+
+- Applying directive fixes after the `verifyAndFix` passes
+  - Not ideal: deleting a comment could introduce new rule violations that would warrant running >=1 other pass
+- Writing an external tool to apply these options
+  - Not ideal: the internal implementation is simple; external would have to deal with variant formatters, etc and suffer from the same introduced rule violations
+
+## Open Questions
+
+- Is `"meta"` a good name for the fix type?
+  - I'd considered `"directive"`, but that felt like it could be too specific to be extended for other meta/configuration in the future.
+
+## Help Needed
+
+I'm looking forward to implementing this if approved! 🙌
+
+## Frequently Asked Questions
+
+> Will these be autofixed by default?
+
+Yes: problems reported by directive usage checking are joined with remaining rule violation problems in a single array.
```

Comment:
I’m not sure having this fixable just when using —fix is the right approach. Right now we don’t fix rules that wouldn’t otherwise output warnings. Since unused directives are not warned by default, I don’t think they should be auto fixed by default. Maybe if both —fix and —report-unused-disable-directives are used, then it would make sense to remove them, or else a separate flag might be a better way to go.

Reply:
Unused directives are indeed warned by default - that’s why shows up in the resulting output.

Reply:
wait, are these not reported by default? i don’t think I’ve ever provided the flag, but I’m pretty sure I’ve seen the output by default

Reply:
I phrased this poorly, and this part of the RFC is inaccurate. This was intended to convey that the fixers for unused directives will be added to reported problems "by default" -- as in, if `--report-unused-disable-directives` is true, there's nothing else you need to do to have the fixers be created. They still won't be applied if `--fix` is not true.

I.e. you must provide _both_ `--fix` and `--report-unused-disable-directives` for them to be fixed.

I'll clarify this wording.

Reply:
Oh, and no, `--report-unused-disable-directives` is IME off by default.

Reply:
Added FAQs in c87e3b9.

Reply:
is it that the flag is required to make them _errors_, but they're _warnings_ by default?

Reply:
It's off by default on the CLI. 

* Default CLI value: https://github.com/eslint/eslint/blob/08ae31e539e381cd0eabf6393fa5c20f1d59125f/conf/default-cli-options.js#L30
* Options processing in the CLI: https://github.com/eslint/eslint/blob/49d16977d969070e5240074e76036f56631a90d3/lib/cli.js#L120
* Defaulting the value to `"off"` in applying disable directives: https://github.com/eslint/eslint/blob/52e2cf50b35d57fb8466e0bcd0581eff1590fb4c/lib/linter/apply-disable-directives.js#L121
* Defaulting the value to `"off"` in verify options: https://github.com/eslint/eslint/blob/9d6063add931f0803cae1676d5df307baf114360/lib/linter/linter.js#L502

Interestingly, CLI-related places in code allow it to be `boolean | undefined` although others allow it to be `"error" | "warn" | "off" | null`. 

I set up a quick little playground for it: https://github.com/JoshuaKGoldberg/eslint-disable-directives-testing

Reply:
There is also [reportUnusedDisableDirectives](https://eslint.org/docs/user-guide/configuring/rules#report-unused-eslint-disable-comments) configuration option.

`reportUnusedDisableDirectives: true` config option produces warnings, `--report-unused-disable-directives` CLI option produces errors.

Reply:
I agree I can't reproduce the behavior I'm describing with your repro - but i have *absolutely* seen yellow warnings about unused disable directives, and i have never ever used the `--report-unused-disable-directives` flag explicitly.

Is there any possibility there's some conditions where the unused disable directive warnings do show up by default?

Reply:
aha! i'm surely using `reportUnusedDisableDirectives: true` in my shared config, and that resolves my confusion :-)

in that case, when that is true in my shared config, they should automatically be autofixed, yes?